<?xml version="1.0"?>

<launch>
  <!-- Topics that need to be chosen by user and can be varied-->
  
    <!-- Whether we want to use nominal / external control or a safety filter -->
  <arg name="safety_filter_active" default="True" />
  
    <!-- Whether we want a single fixed value function (no updating) or updating using hj_reachability_node -->
  <arg name="update_vf_online" default="True" /> 
  
    <!-- What type of initial value function do we have
         This only is relevant if the safety_filter_active = True
         Options: 
          1. cbf:   Load as CBF with parameters set in CBF yaml file
          2. file:  Load as state_discretized value function 
          3. sdf:   Have the signed distance function be used 
    -->
  <arg name="vf_initialization_method" default="sdf" />
  
    <!-- Simulation or hardware -->
  <arg name="hardware" default="True" />

    <!-- What message passing method to use
          Options:
           1. pubsub:   flattens the vf and passes it 
           2. file:     Saves it as a file and pubsubs a Boolean
    -->
  <arg name="vf_update_method" default="pubsub" />

    <!-- Choose which files to load to describe the environment -->
  <arg name="env_config_file" default="simple_env.yaml" />
  
    <!-- Choose which files to load to describe the refinecbf internal nominal controller -->
  <arg name="control_config_file" default="crazyflie_control.yaml" />

    <!-- Choose the name of CBF parameter file or initial vf file -->
  <arg name="CBF_parameter_file" default="crazyflie_CBF_params.yaml" />
  <arg name="initial_vf_file" default="target_values_judy.npy" />

  <!-- For HW experiment, what robot are we dealing with  -->
  <arg name="robot_number" default="1" />

  <!-- ___________________________________________________________________________________________________________ -->
  <!-- Topics that are typically fixed and are shared between different launch files -->
  <arg name="robot_state_topic" value="/state" />
  <arg name="cbf_state_topic" value="/safety_filter/state" />

  <arg name="disturbance_update_topic" value="/env/disturbance_update" />
  <arg name="actuation_update_topic" value="/env/actuation_update" />
  <arg name="sdf_update_topic" value="/env/sdf_update" />
  <arg name="vf_update_topic" value="/safety_filter/vf_update" />

  <arg name="robot_safe_control_topic" value="/robot/final_control" />
  <arg name="cbf_safe_control_topic" value="/safety_filter/safe_control" />
  <arg name="cbf_nominal_control_topic" value="/safety_filter/nominal_control" />
  <arg name="robot_disturbance_topic" value="/robot/disturbance" />
  <arg name="simulated_disturbance_topic" value="/stack/disturbance" />

  <arg name="obstacle_marker_topic" value="/visualization/obstacle_marker" />
  <arg name="sdf_marker_topic" value="/visualization/sdf_marker" />
  <arg name="vf_marker_topic" value="/visualization/vf_marker" />
  <arg name="goal_marker_topic" value ="/visualization/goal" />
  <arg name="obstacle_update_topic" value ="/visualization/obstacle_update" />

  <!-- Services -->
  <arg name="activate_obstacle_service" value="/env/activate_obstacle" />
  
  <!-- Can be changed to accept commands from somewhere in cf stack -->
  <arg name="robot_external_control_topic" value="/robot/external_control" />
  <arg name="cbf_external_control_topic" value="/safety_filter/external_control" />
  <arg name="publish_external_control_flag" value="/safety_filter/externally_controlled" />

  <!-- Loading and storing the environment and nominal control parameters -->
  <arg name="config_file_loc" value="$(find refinecbf_ros)/config/$(arg env_config_file)" />
  <rosparam ns="env" file="$(arg config_file_loc)" command="load" />
  <arg name="control_config_file_loc"
    value="$(find refinecbf_ros)/config/$(arg control_config_file)" />
  <rosparam ns="ctr" file="$(arg control_config_file_loc)" />

  <!-- Nominal control and its bounds -->
  <arg name="K_file" value="$(find crazyflie_lqr)/references/K.txt"/>
  <arg name="u_ref_file" value="$(find crazyflie_lqr)/references/uref$(arg robot_number).txt" if="$(arg hardware)" />
  <arg name="u_ref_file" value="$(find crazyflie_lqr)/references/uref_sim.txt" unless="$(arg hardware)" />
  <arg name="max_thrust" value="15.0" />
  <arg name="min_thrust" value="4.0" />
  <arg name="max_roll" value="0.1" />
  <arg name="max_pitch" value="0.1" />

  <!-- ___________________________________________________________________________________________________________ -->
  <!-- Below: all the nodes being loaded -->
  <node name="cf_nominal_control"
        pkg="refinecbf_ros"
        type="cf_nominal_controller.py"
        output="screen"
        ns="safety_filter">
    <param name="topics/state" value="$(arg cbf_state_topic)" />
    <param name="topics/nominal_control" value="$(arg cbf_nominal_control_topic)" />
    <param name="topics/external_control" value="$(arg cbf_external_control_topic)" />
    <param name="topics/publish_external_control_flag" value="$(arg publish_external_control_flag)" />
    <param name="LQR/K_file" value="$(arg K_file)" />
    <param name="LQR/u_ref_file" value="$(arg u_ref_file)" />
    <param name="control/max_thrust" value="$(arg max_thrust)" />
    <param name="control/min_thrust" value="$(arg min_thrust)" />
    <param name="control/max_roll" value="$(arg max_roll)" />
    <param name="control/max_pitch" value="$(arg max_pitch)" />

  </node> 

  <node name="cf_hw_interface"
        pkg="refinecbf_ros"
        type="cf_hw_interface.py"
        output="screen"
        ns="hw_interface">
    <param name="topics/robot_state" value="$(arg robot_state_topic)" />
    <param name="topics/cbf_state" value="$(arg cbf_state_topic)" />
    <param name="topics/robot_safe_control" value="$(arg robot_safe_control_topic)" />
    <param name="topics/cbf_safe_control" value="$(arg cbf_safe_control_topic)" />
    <param name="topics/robot_external_control" value="$(arg robot_external_control_topic)" />
    <param name="topics/cbf_external_control" value="$(arg cbf_external_control_topic)" />
    <param name="topics/simulated_disturbance" value="$(arg simulated_disturbance_topic)" />
    <param name="topics/robot_disturbance" value="$(arg robot_disturbance_topic)" />
    <param name="control/max_thrust" value="$(arg max_thrust)" />
    <param name="control/min_thrust" value="$(arg min_thrust)" />
    <param name="control/max_roll" value="$(arg max_roll)" />
    <param name="control/max_pitch" value="$(arg max_pitch)" />
  </node>

  <node name="obstacle_node"
    pkg="refinecbf_ros"
    type="obstacle_node.py"
    output="screen">
    <param name="topics/sdf_update" value="$(arg sdf_update_topic)" />
    <param name="topics/cbf_state" value="$(arg cbf_state_topic)" />
    <param name="topics/obstacle_update" value="$(arg obstacle_update_topic)" />
    <param name="services/activate_obstacle" value="$(arg activate_obstacle_service)" />
    <param name="vf_update_method" value="$(arg vf_update_method)" />
  </node>

  <node name="disturbance_node"
    pkg="refinecbf_ros"
    type="disturbance_node.py"
    output="screen">
    <param name="topics/disturbance_update" value="$(arg disturbance_update_topic)" />
    <param name="topics/state" value="$(arg cbf_state_topic)" />
    <param name="topics/disturbance" value="$(arg simulated_disturbance_topic)" />
  </node>

  <node name="crazyflie_visualization"
    pkg="refinecbf_ros"
    type="cf_visualization.py"
    output="screen">
    <param name="topics/sdf_update" value="$(arg sdf_update_topic)" />
    <param name="topics/cbf_state" value="$(arg cbf_state_topic)" />
    <param name="topics/vf_update" value="$(arg vf_update_topic)" />
    <param name="topics/obstacle_update" value="$(arg obstacle_update_topic)" />
    <param name="topics/obstacle_marker" value="$(arg obstacle_marker_topic)" />
    <param name="topics/sdf_marker" value="$(arg sdf_marker_topic)" />
    <param name="topics/vf_marker" value="$(arg vf_marker_topic)" />
    <param name="topics/goal_marker" value="$(arg goal_marker_topic)" />
    <param name="vf_update_method" value="$(arg vf_update_method)" />
  </node>

  <include file="$(find refinecbf_ros)/launch/refine_cbf.launch">
    <arg name="cbf_state_topic" value="$(arg cbf_state_topic)" />
    <arg name="cbf_nominal_control_topic" value="$(arg cbf_nominal_control_topic)" />
    <arg name="cbf_safe_control_topic" value="$(arg cbf_safe_control_topic)" />
    <arg name="vf_update_topic" value="$(arg vf_update_topic)" />
    <arg name="actuation_update_topic" value="$(arg actuation_update_topic)" />
    <arg name="sdf_update_topic" value="$(arg sdf_update_topic)" />
    <arg name="disturbance_update_topic" value="$(arg disturbance_update_topic)" />
    <arg name="safety_filter_active" value="$(arg safety_filter_active)" />
    <arg name="vf_initialization_method" value="$(arg vf_initialization_method)"/>
    <arg name="update_vf_online" value="$(arg update_vf_online)" />
    <arg name="CBF_parameter_file" value="$(arg CBF_parameter_file)" />
    <arg name="initial_vf_file" value="$(arg initial_vf_file)" />
    <arg name="vf_update_method" value="$(arg vf_update_method)" />
  </include>

  <!-- Arguments for child launch files -->
  <!-- Topics. -->

  <arg name="control_reference_topic" value="/control/ref" />
  <arg name="takeoff_control_topic" value="/control/takeoff" />
  <arg name="lqr_control_topic" value="/control/lqr" />
  <arg name="in_flight_topic" value="/control/in_flight" />

  <!-- Frames of reference. -->
  <!-- Fixed frame gets written based on hardware on not hardware -->
  <arg name="robot_frame" value="crazyflie" unless="$(arg hardware)"/>
  <arg name="robot_frame" value="robot_1" if="$(arg hardware)"/>

  <!-- Takeoff hover point. -->
  <arg name="takeoff_hover_x" value="0.0" />
  <arg name="takeoff_hover_y" value="0.0" />
  <arg name="takeoff_hover_z" value="0.8" />

  <arg name="takeoff_thrust" default="13.0" if="$(eval robot_number == 1)" />
  <arg name="takeoff_thrust" default="12.0" if="$(eval robot_number == 2)" />
  <arg name="takeoff_thrust" default="11.5" if="$(eval robot_number == 3)" />
  <arg name="takeoff_duration" default="1.5" />
  <arg name="landing_duration" default="5.0" />

  <!-- Start point for the simulator. NOT SURE IF WE NEED THIS FOR HW EXPERIMENTS, DOESNT HURT TO HAVE THOUGH -->
  <arg name="sim_init_x" value="0.0" />
  <arg name="sim_init_y" value="0.0" />
  <arg name="sim_init_z" value="0.0" />

  <!-- Time steps. -->
  <arg name="simulator_dt" value="0.001" />
  <arg name="estimator_dt" value="0.01" />

  <!-- Dimensions. -->
  <arg name="x_dim" value="7" />
  <arg name="u_dim" value="4" />

  <!-- Frames of reference. -->
  <!-- Fixed frame gets overridden if hardware = True -->
  <arg name="fixed_frame" value="world" unless="$(arg hardware)"/>
  <arg name="fixed_frame" value="odom" if="$(arg hardware)"/>


  <!-- LQR controller. -->
  <node name="lqr_controller"
        ns="/control"
        pkg="crazyflie_lqr"
        type="position_velocity_yaw_state_lift_lqr_node"
        output="screen">

    <param name="x_dim" value="$(arg x_dim)" />
    <param name="u_dim" value="$(arg u_dim)" />

    <param name="K_file" value="$(arg K_file)" />
    <param name="u_ref_file" value="$(arg u_ref_file)" />

    <param name="topics/reference" value="$(arg control_reference_topic)" />
    <param name="topics/state" value="$(arg robot_state_topic)" />
    <param name="topics/control" value="$(arg lqr_control_topic)" />
    <param name="topics/in_flight" value="$(arg in_flight_topic)" />
    <param name="control/max_thrust" value="$(arg max_thrust)" />
    <param name="control/min_thrust" value="$(arg min_thrust)" />
    <param name="control/max_roll" value="$(arg max_roll)" />
    <param name="control/max_pitch" value="$(arg max_pitch)" />
  </node>

  <!-- Takeoff server node. -->
  <node name="takeoff_server"
        ns="/control"
        pkg="crazyflie_takeoff"
        type="takeoff_server_node"
        output="screen">

    <param name="topics/reference" value="$(arg control_reference_topic)" />
    <param name="topics/control" value="$(arg takeoff_control_topic)" />
    <param name="topics/in_flight" value="$(arg in_flight_topic)" />

    <param name="hover/x" value="$(arg takeoff_hover_x)" />
    <param name="hover/y" value="$(arg takeoff_hover_y)" />
    <param name="hover/z" value="$(arg takeoff_hover_z)" />
    <param name="takeoff/thrust" value="$(arg takeoff_thrust)" />
    <param name="duration/open_loop" value="$(arg takeoff_duration)" />
    <param name="duration/landing" value="$(arg landing_duration)" />
  </node>

  <!-- Takeoff control filter node. -->
  <node name="takeoff_control_filter"
        ns="/control"
        pkg="crazyflie_takeoff"
        type="takeoff_control_filter_node"
        output="screen">

    <param name="topics/takeoff_control" value="$(arg takeoff_control_topic)" />
    <param name="topics/commanded_control" value="$(arg lqr_control_topic)" />
    <param name="topics/final_control" value="$(arg robot_external_control_topic)" />
    <param name="topics/in_flight" value="$(arg in_flight_topic)" />
  </node>

  <!-- Launch RViz with the proper config file robot descriptor. -->
  <node name="rviz"
        pkg="rviz"
        type="rviz"
        args="-d $(find crazyflie_simulator)/rviz/crazyflie_simulator.rviz"
        required="true">

    <param name="robot_description"
           command="$(find xacro)/xacro
                    $(find crazyflie_description)/urdf/crazyflie.urdf.xacro" />
  </node>

  <!-- If doing hardware experiments, we need to set addresses etc. -->
  <group if="$(eval arg('hardware') == 1)">
    <!-- Crayzflie addressses -->
    
    <arg name="uri" value="radio://0/80/2M/E7E7E7E7E$(arg robot_number)" />

    <arg name="mocap_frame" value="/natnet_ros/robot_1/pose" />

    <!-- Joystick address -->
    <arg name="joy_dev" default="/dev/input/js0" />

      <!-- TF static publisher to tie the mocap output to the robot frame. -->
    <node pkg="tf" type="static_transform_publisher" name="broadcaster_two" args="0 0 0 0 0 0 1 world odom 100" />
    <node name="cmd_vel_converter"
        pkg="crazyflie_control_merger"
        type="cmd_vel_converter_node"
        output="screen">

      <param name="topics/control" value="$(arg robot_safe_control_topic)" />
      <param name="topics/cmd_vel" value="/$(arg robot_frame)/cmd_vel" />
    </node>

    <!-- Crazyflie interface. -->
    <include file="$(find crazyflie_driver)/launch/crazyflie_server.launch" />

    <group ns="crazyflie">
      <include file="$(find crazyflie_driver)/launch/crazyflie_add.launch">
        <arg name="uri" value="$(arg uri)" />
        <arg name="tf_prefix" value="$(arg robot_frame)" />
        <arg name="enable_logging" value="True" />
      </include>

      <!--
      <node name="joy" pkg="joy" type="joy_node" output="screen">
        <param name="dev" value="$(arg joy_dev)" />
      </node>

      <node name="joystick_controller" pkg="crazyflie_demo" type="controller.py" output="screen">
        <param name="use_crazyflie_controller" value="True" />
      </node>
      -->
    </group>
  </group>
 
  <group if="$(eval arg('hardware') == 0)">

      <!-- Crazyflie dynamics simulator. -->
    <node name="simulator"
        pkg="crazyflie_simulator"
        type="near_hover_simulator_node"
        output="screen">

      <param name="time_step" value="$(arg simulator_dt)" />
      <param name="init/x" value="$(arg sim_init_x)" />
      <param name="init/y" value="$(arg sim_init_y)" />
      <param name="init/z" value="$(arg sim_init_z)" />

      <param name="frames/fixed" value="$(arg fixed_frame)" />
      <param name="frames/robot" value="$(arg robot_frame)" />

      <param name="topics/control" value="$(arg robot_safe_control_topic)" />
      <param name="topics/disturbance" value="$(arg robot_disturbance_topic)" />
    </node>

  </group>

  <!-- State estimator. -->
  <!-- State estimator needs to be instantiated after choosing the fixed frame (sim or real) -->
  <node name="state_estimator"
        pkg="crazyflie_state_estimator"
        type="position_velocity_yaw_state_estimator_node"
        output="screen">

    <param name="x_dim" value="$(arg x_dim)" />
    <param name="time_step" value="$(arg estimator_dt)" />

    <param name="frames/fixed" value="$(arg fixed_frame)" />
    <param name="frames/robot" value="$(arg robot_frame)" />

    <param name="topics/state" value="$(arg robot_state_topic)" />
  </node>

</launch>
